wrapper/source_hosts.py

```diff
@@ -0,0 +1,773 @@
+                name='rhosp-migration-{}'.format(volume_id),
+                timeout=DEFAULT_TIMEOUT)
+
+            # Create a new volume from the snapshot
+            logging.info('Creating new volume from root snapshot')
+            root_volume_copy = self.conn.create_volume(
+                wait=True, name='rhosp-migration-{}'.format(volume_id),
+                snapshot_id=root_snapshot.id, size=root_snapshot.size,
+                timeout=DEFAULT_TIMEOUT)
+
+            # Update the volume map with the new volume ID
+            self.volume_map['/dev/vda'] = mapping._replace(
+                source_id=root_volume_copy.id,
+                snap_id=root_snapshot.id)
+        elif sourcevm.image:
+            logging.info('Image-based instance, creating snapshot...')
+            image = self.conn.compute.create_server_image(
+                name='rhosp-migration-root-{}'.format(sourcevm.name),
+                server=sourcevm.id)
+            for second in range(DEFAULT_TIMEOUT):
+                refreshed_image = self.conn.get_image_by_id(image.id)
+                if refreshed_image.status == 'active':
+                    break
+                time.sleep(1)
+            else:
+                raise RuntimeError(
+                    'Could not create new image of image-based instance!')
+            volume = self.conn.create_volume(
+                image=image.id, bootable=True, wait=True, name=image.name,
+                timeout=DEFAULT_TIMEOUT, size=image.min_disk)
+            disk = Disk('/dev/vda', 0)
+            self.volume_map['/dev/vda'] = VolumeMapping(
+                source_dev=None, source_id=volume.id, dest_dev=None,
+                dest_id=None, snap_id=None, image_id=image.id,
+                name=volume.name, size=volume.size, url=None, state=disk)
+            STATE.disks.append(disk)
+            logging.debug('STATE.disks is now %s', STATE.disks)
+            STATE.write()
+        else:
+            raise RuntimeError('No known boot device found for this instance!')
+
+        for path, mapping in self.volume_map.items():
+            if path != '/dev/vda':  # Detach non-root volumes
+                volume_id = mapping.source_id
+                volume = self.conn.get_volume_by_id(volume_id)
+                logging.info('Detaching %s from %s', volume.id, sourcevm.id)
+                self.conn.detach_volume(server=sourcevm, volume=volume,
+                                        wait=True, timeout=DEFAULT_TIMEOUT)
+
+    def _wait_for_volume_dev_path(self, conn, volume, vm, timeout):
+        volume_id = volume.id
+        for second in range(timeout):
+            volume = conn.get_volume_by_id(volume_id)
+            if volume.attachments:
+                attachment = self._get_attachment(volume, vm)
+                if attachment.device.startswith('/dev/'):
+                    return
+            time.sleep(1)
+        raise RuntimeError('Timed out waiting for volume device path!')
```

Comment:
So this (and other functions down from here) is why I wanted someone who is going to be working on openstack2openstack to have a look at the two-phase implementation for Openstack. Some of these already exist in the codebase, but are poorly written (by me, so really poorly ;)) using the openstack cli.  It would be nice to:
 1) use the API instead
 2) not duplicate the code